core/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -148,7 +148,14 @@ public MapperService(IndexSettings indexSettings, IndexAnalyzers indexAnalyzers,
         this.searchQuoteAnalyzer = new MapperAnalyzerWrapper(indexAnalyzers.getDefaultSearchQuoteAnalyzer(), p -> p.searchQuoteAnalyzer());
         this.mapperRegistry = mapperRegistry;
 
-        this.dynamic = this.indexSettings.getValue(INDEX_MAPPER_DYNAMIC_SETTING);
+        if (indexSettings.getIndexVersionCreated().onOrAfter(Version.V_6_0_0_alpha1)) {
+            if (INDEX_MAPPER_DYNAMIC_SETTING.exists(indexSettings.getSettings())) {
+                throw new IllegalArgumentException("Setting " + INDEX_MAPPER_DYNAMIC_SETTING.getKey() + " was removed after version 6.0.0");
+            }
+            this.dynamic = INDEX_MAPPER_DYNAMIC_DEFAULT;
+        } else {
+            this.dynamic = this.indexSettings.getValue(INDEX_MAPPER_DYNAMIC_SETTING);
+        }
```

Comment:
master is the future 7.0, so I would do the following:
```java
if (INDEX_MAPPER_DYNAMIC_SETTING.exists(indexSettings.getSettings())) {
    if (indexSettings.getIndexVersionCreated().onOrAfter(Version.V_7_0_0)) {
        // 7.x index
        throw new IllegalArgumentException("Setting " + INDEX_MAPPER_DYNAMIC_SETTING.getKey() + " was removed after version 6.0.0");
    } else {
        // 6.x index
        DEPRECATION_LOGGER.deprecated("Setting " + INDEX_MAPPER_DYNAMIC_SETTING.getKey() + " is deprecated since indices may not have more than one type anymore.");
    }
}
```

Then when backporting I'll just remove the 7.x branch and make sure that we only emit a deprecation warning on 6.x indices (you don't need to worry about it).